aptos-move/framework/aptos-framework/sources/state_storage.move

```diff
@@ -1,59 +1,82 @@
 module aptos_framework::state_storage {
 
-    use aptos_framework::reconfiguration;
     use aptos_framework::system_addresses;
     use std::error;
 
-    friend aptos_framework::genesis;
     friend aptos_framework::block;
+    friend aptos_framework::genesis;
+    friend aptos_framework::reconfiguration;
 
-    const EUSAGE_ALREADY_EXISTS: u64 = 0;
-    const EEPOCH_ZERO: u64 = 1;
+    const ESTATE_STORAGE_USAGE: u64 = 0;
+    const EGAS_PARAMETER: u64 = 1;
+    const EEPOCH_ZERO: u64 = 2;
 
-    struct StateStorageUsage has copy, drop, key, store {
-        epoch: u64,
+    struct Usage has copy, drop, store {
         items: u64,
         bytes: u64,
     }
 
-    struct Usage_ {
-        items: u64,
-        bytes: u64,
+    /// This is updated at the begining of each opoch, reflecting the storage
+    /// usage after the last txn of the previous epoch is committed.
+    struct StateStorageUsage has key, store {
```

Comment:
If I understand correctly, you write to StateStorageUsage in the first block of the new epoch and to GasParameter in the last block of the previous epoch. This seems redundant. Why do we need to do this?

Reply:
> you write to StateStorageUsage in the first block of the new epoch

the comment on `native fun get_state_storage_usage_only_at_epoch_beginning()` -- the only occasion every node is guaranteed to be able to read it

> and to GasParameter in the last block of the previous epoch

the comment on `struct GasParameter has key, store` -- the vm needs the value to not change except in the last txn in an epoch

Reply:
Talked to Alden offline and this makes sense now 👍